In [100]:
import pandas as pd
import numpy as np
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score
from DecisionTree import TreeNode,DecisionTree

In [101]:
trainDataset=pd.read_csv("train.csv")
testDataset=pd.read_csv("test.csv")
trainDataset.head()

ID     Date     Opponent Is_Home_or_Away Is_Opponent_in_AP25_Preseason  \
0   1   9/5/15        Texas            Home                           Out   
1   2  9/12/15     Virginia            Away                           Out   
2   3  9/19/15  GeorgiaTech            Home                            In   
3   4  9/26/15        UMass            Home                           Out   
4   5  10/3/15      Clemson            Away                            In   

   Media Label  
0  1-NBC   Win  
1  4-ABC   Win  
2  1-NBC   Win  
3  1-NBC   Win  
4  4-ABC  Lose

In [102]:
trainDataset=trainDataset.drop(["Opponent","Date","ID"], axis=1)
testDataset=testDataset.drop(["Opponent","Date","ID"], axis=1)
trainDataset.head()

Is_Home_or_Away Is_Opponent_in_AP25_Preseason  Media Label
0            Home                           Out  1-NBC   Win
1            Away                           Out  4-ABC   Win
2            Home                            In  1-NBC   Win
3            Home                           Out  1-NBC   Win
4            Away                            In  4-ABC  Lose

In [103]:
#Categorical data => numerical trainDataset
for col in trainDataset.columns:
    trainDataset[col]=pd.Categorical(trainDataset[col])
    trainDataset[col]=trainDataset[col].cat.codes
    
trainDataset

Is_Home_or_Away  Is_Opponent_in_AP25_Preseason  Media  Label
0                 1                              1      0      1
1                 0                              1      3      1
2                 1                              0      0      1
3                 1                              1      0      1
4                 0                              0      3      0
5                 1                              1      0      1
6                 1                              0      0      1
7                 0                              1      3      1
8                 0                              1      3      1
9                 1                              1      0      1
10                0                              1      0      1
11                0                              0      2      0
12                0                              1      3      0
13                1                              1      0      1
14                1                              1      0      0
15                1                              1      0      0
16                1                              1      1      1
17                0                              1      3      0
18                1                              0      0      0
19                1                              1      0      1
20                1                              1      4      0
21                1                              1      0      1
22                1                              0      0      0
23                0                              0      3      0

In [104]:
#Categorical data => numerical testDataset
for col in testDataset.columns:
    testDataset[col]=pd.Categorical(testDataset[col])
    testDataset[col]=testDataset[col].cat.codes
testDataset

Is_Home_or_Away  Is_Opponent_in_AP25_Preseason  Media  Label
0                 1                              1      0      1
1                 1                              0      0      0
2                 0                              1      1      1
3                 0                              1      2      1
4                 1                              1      0      1
5                 0                              1      3      1
6                 1                              1      0      1
7                 1                              1      0      1
8                 1                              1      0      1
9                 0                              0      3      0
10                1                              1      0      1
11                0                              0      3      0

In [105]:
#spearting labels and features
X= trainDataset[trainDataset.columns[:-1]].values
Y=trainDataset['Label'].values
X

array([[1, 1, 0],
       [0, 1, 3],
       [1, 0, 0],
       [1, 1, 0],
       [0, 0, 3],
       [1, 1, 0],
       [1, 0, 0],
       [0, 1, 3],
       [0, 1, 3],
       [1, 1, 0],
       [0, 1, 0],
       [0, 0, 2],
       [0, 1, 3],
       [1, 1, 0],
       [1, 1, 0],
       [1, 1, 0],
       [1, 1, 1],
       [0, 1, 3],
       [1, 0, 0],
       [1, 1, 0],
       [1, 1, 4],
       [1, 1, 0],
       [1, 0, 0],
       [0, 0, 3]], dtype=int8)

In [106]:
predFeatureX=testDataset[testDataset.columns[:-1]].values
xGroundTruth=testDataset["Label"].values
predFeatureX


array([[1, 1, 0],
       [1, 0, 0],
       [0, 1, 1],
       [0, 1, 2],
       [1, 1, 0],
       [0, 1, 3],
       [1, 1, 0],
       [1, 1, 0],
       [1, 1, 0],
       [0, 0, 3],
       [1, 1, 0],
       [0, 0, 3]], dtype=int8)

In [107]:
xGroundTruth

array([1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0], dtype=int8)

In [108]:
#Naive baiyes prediction

model=GaussianNB()
model.fit(np.array(X),Y)
predY=model.predict(predFeatureX)

## Question 2

Write down prediction labelsof the 12 testing games in the PDF.

In [109]:
predY

array([1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0], dtype=int8)

## Ans: - Win, Lose, Win, Win, Win, Lose, Win, Win, Win, Lose, win, Lose

In [110]:
# Calculation of accuracy and fscore

def accuracyCheck(name, y, y_pred):
    headers = {
        'Accuracy': accuracy_score, 
        'F1_Score': f1_score, 
        'Precision': precision_score, 
        'Recall': recall_score
    }
    
    with open('{}.csv'.format(name), 'w') as summary:
        print(','.join(headers.keys()), file=summary)
        summary.write(','.join(['{:3.3f}'.format(metric(y, y_pred)) for metric in headers.values()]))

## Question 1

Programming: Use Naïve Bayes to predict labels of instances in the testing set (12 games) based on the training set (24 games). Calculate Accuracy, Precision, Recall, andF1 score on the testing result.

In [111]:
accuracyCheck("NaiveBayesPred",xGroundTruth,predY)
NaiveBayesAcc=pd.read_csv("NaiveBayesPred.csv")
NaiveBayesAcc

Accuracy  F1_Score  Precision  Recall
0     0.917     0.941        1.0   0.889

## Ans: -
## Accuracy of Naive Bayes - 0.917
## F1 Score of Navie Bayes - 0.941
## Precision of Navie Bayes - 1.0
## Recall of Naive Bayes - 0.889

## Question 3

Compare Naïve Bayes with ID3 and C4.5, which model is the best, which model performs the worst? Can you explain why?

Ans: - 
We have calculated F1 score, accuracy, precision and recall by c4.5 and ID3 on same dataset in last assignment. We got following result: -

ID3 Accuracy - 0.833
ID3 F1 Score - 0.889
ID3 Precision - 0.889
ID3 Recall - 0.889

C4.5 Accuracy - 0.917
C4.5 F1 Score - 0.947
C4.5 Precision - 0.9
C4.5 Recall - 1.0

and from above, we have calculated same measures using Naive Bayes Model. And their values are as follow: -

Navie Bayes Accuracy: - 0.917
Navie Bayes F1 Score: - 0.941
Navie Bayes Precision: - 1.0
Navie Bayes Recall: - 0.889

From above observations, we can say that Naive Bayes and C4.5 models are almost similar. But Naive Bayes(NB) works better than C4.5 as Navie Bayes precision is 1 and, C4.5 precision is 0.9. 

As Naive Bayes works better for this dataset on other hand ID3 model performs poor as compared to Naive Bayes and C4.5 model.

Naive Bayes works better with smaller dataset. Smaller dataset means faster processing for Naive Bayes.
In Decision Tree(ID3) is over fitied if there is a smaller dataset. Classifying Continuous data is computationally expensive, as many trees must be generted to see where to break the continuum.


Colab Link---------------- https://colab.research.google.com/drive/1KWfgaUzqiy2DvCtnf_wEc5LTwl8kMBMt